Plugins/Events/Events/SpellEvents.cpp

```diff
@@ -249,5 +286,155 @@ int32_t DecrementSpellReadyCountHook(CNWSCreature *thisPtr, uint32_t nSpellID, u
     return retVal;
 }
 
+uint32_t AIActionCastSpellHook(CNWSCreature* thisPtr, CNWSObjectActionNode* pNode)
+{
+    LastSpellAction.nSpellId = pNode->m_pParameter[0];
+    LastSpellAction.nMultiClass = pNode->m_pParameter[1] & 0xFF;
+    LastSpellAction.nDomainLevel = pNode->m_pParameter[2];
+    LastSpellAction.nMetaMagic = pNode->m_pParameter[3];
+    LastSpellAction.bSpontaneous = pNode->m_pParameter[4] != 0;
+    LastSpellAction.bDefensiveCast = (pNode->m_pParameter[1] & 0x0000FF00) >> 8;
+    LastSpellAction.oidTarget = pNode->m_pParameter[5];
+    LastSpellAction.fTargetX = *((float*)&pNode->m_pParameter[6]);
+    LastSpellAction.fTargetY = *((float*)&pNode->m_pParameter[7]);
+    LastSpellAction.fTargetZ = *((float*)&pNode->m_pParameter[8]);
+    LastSpellAction.bInstant = (pNode->m_pParameter[9] & 0x40000000) != 0;
+    LastSpellAction.nProjectilePathType = pNode->m_pParameter[9] & 0x3FFFFFFF;
+    LastSpellAction.nFeat = pNode->m_pParameter[10];
+    LastSpellAction.nCasterLevel = pNode->m_pParameter[11] & 0x000000FF;
+    LastSpellAction.bFake = (pNode->m_pParameter[9] & 0x80000000) != 0;
+
+    return s_AIActionCastSpellHook->CallOriginal<uint32_t>(thisPtr, pNode);
+}
+
+BOOL ClearActionHook(CNWSCreature* thisPtr, CNWSObjectActionNode* pNode, BOOL bIsTopmostAction)
+{
+    if ((IsIDInWhitelist("NWNX_ON_SPELL_FAILED", pNode->m_pParameter[0])) && 
```

Comment:
You check the whitelist, but I don't see any info in the documentation about it using it.

The whitelist isn't enabled by default, I'm not sure if that's what you intended, you can force enable it by calling `ForceEnableWhitelist("NWNX_ON_SPELL_FAILED");` in the `InitOnFirstSubscribe()` bit if needed

Reply:
I was going for the same behaviour as the other spell events which are using a whitelist that's not enabled by default. Do you think it should be whitelisted by default?